rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,461 @@
+
+```tsx
+<Box
+  display="flex"
+  flexDirection="column"
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+  gap="spacing.1"
+>
+  <Button>Button 1</Button>
+  <Button>Button 2</Button>
+  <Button>Button 3</Button>
+</Box>
+```
+
+Since blade is not a general purpose design system we should lean more towards constraints initially, assess usecases in the future and then increase flexibility. Proposed props for initial version:
+
+| Category            | Props                                       | Notes                                                                                                                                  |
+| ------------------- | ------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| Spacing             | `margin*`, `padding*`,                      | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#space) referenced from styled-system.         |
+| Layout              | `width*`, `height*`, `overflow*`, `display` | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#layout) referenced from styled-system.        |
+| Flexbox             | `flex*`, `align*`, `justify*`, `order`      | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#flexbox) referenced from styled-system.       |
+| Gap                 | `gap`, `rowGap`, `columnGap`                | Provides native flex functionality for spacing siblings.                                                                               |
+| Position            | `position`, `zIndex`, `top`, etc.           | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#position) referenced from styled-system.      |
+| Grid **(web only)** | `grid*`                                     | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#grid-layout)\* referenced from styled-system. |
+
+> **Note**
+>
+> Not all CSS grid properties are covered in `styled-system` spec. These should be referenced from MDN during implementation. Here's a nice reference from [CSS Tricks](https://css-tricks.com/snippets/css/complete-guide-grid).
+
+**Spacing values**
+
+- All spacing based properties should accept spacing tokens i.e. `spacing.1`, `spacing.2`, etc.
+- TypeScript auto suggestion hints should also show the `px` value associated with a spacing token, this helps in translating design from Figma to code since designs would have these in `px`.
+- They should also accept `number` or `string` values eg. `1`, `'1px'`, `'100%'`, `'auto'`, `'1fr'` etc. Numeric literals imply the pixel value (as in vanilla React), if someone wants to use a spacing token it must be prefixed like `spacing.*`.
```

Comment:
imo we shouldn't accept unitless values like `1`. If the consumers are choosing to not rely on our tokens and pass their custom value, it would be better if they also specify the unit along with it instead of us appending a unit to it later on. 

The only case where we should do any logical computation here is when a consumer passes a spacing token, otherwise take the value as is

Reply:
React [automatically transforms](https://reactjs.org/docs/dom-elements.html#style) unitless values to pixel values, so I thought it might be good to keep this closer to that. A related downside is we'll have to add runtime checks to throw in case of numeric values (for JS projects). But yes, it does have the downside which I also added in drawbacks - leading to confusions _does this 1 mean 1 px or `spacing.1`_.

Reply:
+1

Reply:
Updated spec to disallow `number` values in https://github.com/razorpay/blade/pull/950/commits/f4e3b6cebb79e8d1a3b72ed54eb53ed67c55dec4